### PR TITLE
[Fix #3335] Mass assignment in members controller

### PR DESCRIPTION
**Pull request by machisuji**
[WP #3335](https://www.openproject.org/work_packages/3335)

### CODE REVIEW

File: app/controllers/members_controller.rb [hunk 1]
```diff
@@ -28,6 +28,8 @@
 #++
 
 class MembersController < ApplicationController
+  include ActiveModel::ForbiddenAttributesProtection
```
Comment (tessi) [inline app/controllers/members_controller.rb]: this should be included in the model, not in the controller :)

Comment (machisuji) [inline app/controllers/members_controller.rb]: Good point. Interesting that it still works, though ...

File: app/controllers/members_controller.rb [hunk 2]
```diff
@@ -148,13 +150,12 @@ def autocomplete_for_member
 
   def new_members_from_params
     members = []
-
     attrs = params[:member].dup
     user_ids = possibly_seperated_ids_for_entity(attrs, :user)
     roles = Role.find_all_by_id(possibly_seperated_ids_for_entity(attrs, :role))
```
Comment (machisuji) [inline app/controllers/members_controller.rb]: I wonder if this actually ever happens (role and/or user IDs being transmitted in a single field as a comma-separated list).
If it doesn't, we could make that code slightly shorter as well.

**Comment by tessi** [comment]
Hrm.. the features/members/error_messages.feature:83 seem to be failing

File: app/controllers/members_controller.rb (hunk 2 above)
Comment (tessi) [inline app/controllers/members_controller.rb]: I'm not sure, but I could play a little and inspect the params hash for that `:role` attribute

File: app/controllers/members_controller.rb (hunk 1 above)
Comment (tessi) [inline app/controllers/members_controller.rb]: It "worked" because `ForbiddenAttributesProtection` thows an error when mass assignment is used with non-whitelisted parameters. When you include that module in the controller, the model is still unprotected (and does not complain). And you never mass assign attributes to the controller, don't you? :)

**Comment by tessi** [comment]
Also there is more mass assignment on members than one might think. For example [UsersController#edit_membership](https://github.com/opf/openproject/blob/dev/app/controllers/users_controller.rb#L242) mass assigns params. There might be even more places where this folly happens.

**Comment by ulferts** [comment]
Sorry for stating the obvious but one test still fails (I want to generate a mail that will lead to the test being fixed, the PR reevaluated, merged, the OP ticket closed and the world a better place.)

Did you have the chance to try your PR with our plugins, especially impermanent-memberships?

File: spec/controllers/members_controller_spec.rb [hunk 3]
```diff
@@ -40,6 +40,92 @@
...
+
+      response.should_not be_success
+      Member.find(member_2.id).attributes[field.to_s].should_not == value
+    end
+
+    before do
+      User.stub(:current).and_return(admin)
+    end
+
+    it "should specifically not allow 'user_id' to be mass assigned" do
```
Comment (ulferts) [inline spec/controllers/members_controller_spec.rb]: We typically spec mass assignment in permitted_params_spec.rb.

It has the advantage of being faster and there are also less concerns in the controller specs. If you want to ensure that permitted params is called, you could always use should_receive.

File: app/controllers/members_controller.rb [hunk 4]
```diff
@@ -148,32 +148,33 @@ def autocomplete_for_member
 
   def new_members_from_params
     members = []
+    user_ids = possibly_seperated_ids_for_entity(params[:member], :user)
+    roles = Role.find_all_by_id(possibly_seperated_ids_for_entity(params[:member], :role))
 
-    attrs = params[:member].dup
-    user_ids = possibly_seperated_ids_for_entity(attrs, :user)
-    roles = Role.find_all_by_id(possibly_seperated_ids_for_entity(attrs, :role))
+    new_member = lambda do |user_id|
```
Comment (ulferts) [inline app/controllers/members_controller.rb:153]: You sir, are a functional guy.

File: app/controllers/members_controller.rb [hunk 5]
```diff
@@ -148,32 +148,33 @@ def autocomplete_for_member
...
+      end
+    end
 
     user_ids.each do |user_id|
-      member = Member.new attrs
-      # workaround due to mass-assignment protected member_roles.role_id
-      member.member_roles << roles.collect {|r| MemberRole.new :role => r }
-      member.user_id = user_id
-      members << member
+      members << new_member.call(user_id)
```
Comment (ulferts) [inline app/controllers/members_controller.rb]: you could shorten this to 

<pre>
members = user_ids.map { |user_id| new_member.call(user_id) }
</pre>


Then, you would be able to remove line 150. The necessary adaptation of 164ff shouldn't be to hard.

Just a suggestion, however.

**Comment by ulferts** [comment]
**Please do not merge this till the plugins dependent on this are also adapted.**

**Comment by ulferts** [comment]
Hey @machisuji,

once you have finished your current work, can you please see to the plugins so we can merge this.

**Comment by machisuji** [comment]
Will do!

**Comment by machisuji** [comment]
If this is merged [that](https://github.com/finnlabs/openproject-dtag_customizing/issues/10) will have to be merged as well, otherwise the respective plugin breaks stuff.

**Comment by machisuji** [comment]
Looks like we have a toppler here. Don't worry! Everything should be ok!
I obviously also ran the tests in conjunction with the plugin for which the pull request from above is.

File: app/controllers/members_controller.rb [hunk 6]
```diff
@@ -147,33 +147,33 @@ def autocomplete_for_member
...
-      member.user_id = user_id
-      members << member
+    user_ids = possibly_seperated_ids_for_entity(params[:member], :user)
+    roles = Role.find_all_by_id(possibly_seperated_ids_for_entity(params[:member], :role))
+
+    new_member = lambda do |user_id|
+      Member.new(permitted_params.member).tap do |member|
+        member.user_id = user_id if user_id
+      end
+    end
```
Comment (hschink) [inline app/controllers/members_controller.rb:157]: Correct me, but shouldn't this be a non-anonymous method like `create_member_for_user` or something similar? In other words: I don't know why we need an anonymous method here.
